### ISSUE TITLE
3.0 - Product Requirement flat rate discount amount is incorrect when scope is specified

### ISSUE THREAD

**Issue by davidsherlock** [easydigitaldownloads/easy-digital-downloads#8757]
### Expected behavior
I expect flat rate discount amounts to be correctly distributed across all `exclusive` cart items when specific rather than applied directly to the amount zeroing it out. This happens when non-exclusive, but not exclusive as the amount value is just passed to the discount class method for basic calculation.

### Actual behavior
Flattens amounts rather than distributes the flat rate amount across the exclusive items.

### Steps to reproduce the behavior
1) Create a discount, and add product requirements for 2-3 products. Set condition to all and scope to exclusive.
2) Make it a flat rate amount that is generally higher than the product prices.
3) Do not add excluded products and add the items to the cart, you can add additional items too.
4) Apply the discount and you will see the flat rate amount is not distributed, but applied on a per-item basis.

### Information (if a specific version is affected):
PHP Version: 8.0

EDD Version (or branch): 3.0-beta3

WordPress Version: latest

**Comment by davidsherlock** [comment]
This can essentially be fixed by forking the else condition logic and moving it to a separate function to keep things DRY, then passing in an array of download ID and cart items to discount. This will then run the logic but only on the exclusive items, keeping the amounts distributed. This would obviously need to exclude products, etc.

**Comment by davidsherlock** [comment]
This actually results in the customer getting significantly more off also, see this screenshot for example:

<img width="681" alt="Screenshot 2021-07-15 at 18 53 55" src="https://user-images.githubusercontent.com/22668877/125818634-65c0c2f3-2efe-4e30-8f77-8e61c2df5850.png">

This was a $10 off flat rate discount, but $20 has been applied because it has applied the $10 to both items.

**Comment by ashleyfae** [comment]
Is this definitely different behaviour in 3.0 than master branch?

**Comment by davidsherlock** [comment]
I haven't tested with the older one, but I noticed most of the functions were re-factored for 3.0 so assumed it is the same logically speaking.

I can test it tomorrow and report back if you want?

**Comment by ashleyfae** [comment]
Yes please. :+1: If this is also the behaviour in master then I'm not sure if we'll change it. (Or at least it definitely won't be changed for 3.0 and we'd have to carefully consider it for later.) It's quite significant to go from a flat discount across _all items_ versus a flat discount on the cart.

**Comment by davidsherlock** [comment]
Yeah, I agree. For now it's fine, but worth mentioning because it's not "expected" behaviour per-say...

I will do some tests and see. In the meantime, @ashleyfae could you check out this issue #8761 and let me know your thoughts. Would be helpful to have this merged, as working on something discount related :)

**Comment by davidsherlock** [comment]
This isn't super pretty, but it fixes the issue:

```
$items_amount = 0;

if ( ! empty( $reqs ) && 'global' !== $discount->get_scope() ) {
  $items = $reqs;
}

foreach ( $items as $i ) {
  if ( ! in_array( $i['id'], $excluded_products ) ) {
    if ( edd_has_variable_prices( $i['id'] ) ) {
      $i_amount = edd_get_price_option_amount( $i['id'], $i['options']['price_id'] );
    } else {
      $i_amount = edd_get_download_price( $i['id'] );
    }

    $items_amount += ( $i_amount * $i['quantity'] );
  }
}
```

**Comment by robincornett** [comment]
@davidsherlock have you been able to confirm that flat discounts are definitely behaving differently in 3.0 than in 2.11.x?

**Comment by davidsherlock** [comment]
I have not tested, but the method was only introduced in 3.0 so _maybe_...

**Comment by robincornett** [comment]
The cart calculations do not use the new `edd_validate_discount` function--that serves a slightly different purpose. It looks like the cart calculations use the `edd_get_item_discount_amount` function in both 2.x and 3.0, and I think I'm actually seeing this flat discount duplication in 2.x. It does seem like a bug to me, but not one that we've introduced in 3.0 that I can see.

Here's my cart in 2.11.x:
![image](https://user-images.githubusercontent.com/4853066/157679869-e9f1c7aa-8f88-48d9-9825-711ac3baabd3.png)

I created a flat discount for $10, set the product requirements to these two products, set the scope to "apply discount only to selected downloads", and set the product condition to "all". It seems like the $10 discount should be applied across the two products, $5 each, but it is being doubled, and I do suspect that's not right. Definitely keeping this open to discuss, but not milestoning for 3.0 as it does not appear to be a new issue.

**Comment by davidsherlock** [comment]
The issue is rooted in that function, but I can fix it, just need some time to make it nice for EDD code base :)

**Comment by davidsherlock** [comment]
The primary issue is that function needs to be broken down into sub functions, to keep the code DRY. The issue is the first IF condition which does not account for distribution of flat rate amounts. All the code in the `else` needs to be moved to a sub function and then called directly. The way this entire function is written is pretty 'messy' so @ashleyfae might be right about leaving it for now...

**Comment by robincornett** [comment]
I think the bulk of that function was done to get something similar to `edd_validate_discount` -- to get a function which could be used independently of the cart. It looks like it's more or less the same as the original `get_item_discount_amount` method in the current cart class, though (in 3.0 that method now calls `edd_get_item_discount_amount` after normalizing some parameters).

Definitely complicated and worth looking into, but I believe it's consistent with the current behavior, so not marking as urgent for now.
